src/Microsoft.AspNetCore.SignalR.Core/HubGroupList.cs

```diff
@@ -0,0 +1,96 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Collections;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Microsoft.AspNetCore.SignalR
+{
+    public class HubGroupList : IReadOnlyCollection<ConcurrentDictionary<string, HubConnectionContext>>
+    {
+        private readonly ConcurrentDictionary<string, GroupConnectionList> _groups =
+            new ConcurrentDictionary<string, GroupConnectionList>();
+
+        private static readonly GroupConnectionList EmptyGroupConnectionList = new GroupConnectionList();
+
+        public ConcurrentDictionary<string, HubConnectionContext> this[string groupName]
+        {
+            get
+            {
+                _groups.TryGetValue(groupName, out var group);
+                return group;
+            }
+        }
+
+        public void Add(HubConnectionContext connection, string groupName)
+        {
+            CreateOrUpdateGroupWithConnection(groupName, connection);
+        }
+
+        public void Remove(string connectionId, string groupName)
+        {
+            if (!_groups.TryGetValue(groupName, out var connections)) return;
+            ICollection<KeyValuePair<string, GroupConnectionList>> col = _groups;
+            if (!connections.TryRemove(connectionId, out var _) || !connections.IsEmpty) return;
+            var groupToRemove =
```

Comment:
Isn't it racey though? What happens if there is a context switch at this point and a connection was added to the group. Now we go back to this thread and will remove the group even though it actually has some connections.

Reply:
No, because we check for both, key and value, and if added new connection to group from another thread we don't remove the group.

Reply:
@Gurgen  - Sorry, you are right. @BrennanConroy just pointed that we are using kvp and therefore it will work correctly.

Reply:
Ok,, I prepare new commit, please check